### PR TITLE
Set type="password" on all auth_token fields

### LINKED ISSUE #22174: Migrating: Auth token is not properly protected in frontend
## Description

The migration form exposes the auth token to screen capture/cameras/eyeballs.

Browsers also pick this up, adding it to their auto complete dictionary.

*Note:* I already have a fix for this, I just wanted an issue to reference.

### Gitea Version

27746715842da4739d3dba2f3c423df520113a18

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

exposed auth token

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

Reproducible on try.gitea.io, so however that's running.

### Database

None

# Notes from other discussions:

> Yeah, hiding it is IMHO just asking for additional whitespaces causing troubles,... I'd also tend to just prevent auto-completion.
>
> -- _Originally posted by gapodo in https://codeberg.org/forgejo/forgejo/issues/150#issuecomment-732244_

I see this as needing replacement with a password input with an eyeball to hide/show the password. E.g.:

password-show-hide-for-login-form

But I'm not sure what the best way to implement that in the current Gitea project is (or if the input type is already implemented elsewhere and I just need to pull it in). Anything I'd do, I'd want to make reusable.

### PR DESCRIPTION
**Pull request by h3xx**
Set `type="password"` on all `auth_token` fields

Seen when migrating from other hosting platforms.

1. Prevents exposing the token to screen capture/cameras/eyeballs.
2. Prevents the browser from saving the value in its autocomplete dictionary, which often is not secure.

exposed auth token

Closes #22174

### CODE REVIEW

**Comment by delvh** [comment]
Well… This approach has the disadvantage that the browser will ask if the "password" can be saved, which is not a good idea either.
There's no middle ground to hide the text like a password without being treated as a password, right?
I'm not sure if I prefer the proposed approach or the current one.
I think I'm neutral on this PR.

**Comment by silverwind** [comment]
> Well… This approach has the disadvantage that the browser will ask if the "password" can be saved, which is not a good idea either.

Set [autocomplete="new-password"](https://developer.mozilla.org/en-US/docs/Web/Security/Securing_your_site/Turning_off_form_autocompletion#preventing_autofilling_with_autocompletenew-password) to prevent autofill. This is honored by all password managers as far as I know. Not ideal, but many password managers ignore `autocomplete=off`, so this is the best way currently.

**Comment by silverwind** [comment]
To prevent autofill, one method is to change `type` from "text" to "password" after the first `input` event in JS. Should check if such a method would also prevent browser save dialogs.

**Comment by silverwind** [comment]
How do other sites handle such an issue? Any examples where auth token input is masked without the downsides of a password field?

**Comment by silverwind** [comment]
There does exist a [font hack](https://stackoverflow.com/a/22457652/808699). I guess this may be acceptable if the font is lightweight (ideally woff2) and can ideally be embedded as base64 in CSS.

**Comment by silverwind** [comment]
Here is a possible hack that replaces the font with dots. I think the only downside is that users may not see platform-native dots, but the ones provided by the font.

https://jsfiddle.net/silverwind/y4nfo9sx/

**Comment by silverwind** [comment]
https://github.com/go-gitea/gitea/pull/22232 with the suggested approach.

**Comment by silverwind** [comment]
Approach is too much of a hack, so I will not chase it further. To quote myself:

> Proper solution should be type=password, we just need to figure out how to suppress the browser from offering a "Save Password" prompt. Some other forms like webhook already use type=password and there is no prompt, so it must be possible somehow.

**Comment by silverwind** [comment]
Maybe `autocomplete=off` will magically remove those prompts, I see we use it on some other password fields as well.

**Comment by silverwind** [comment]
Sounds good let's add it. Could you verify in at least Chrome and Firefox? If it works there, I think we are good to merge this.

**Comment by h3xx** [comment]
I got to thinking, "how do other sites construct similar forms?" Gitlab has a similar-looking one at `https://gitlab.com/<user>/<project>/-/hooks`:

image

That form is constructed with URL being `type="text" autocomplete="off"`, and the Token input is `autocomplete="new-password" type="password"`.

<details>
<summary>(click to expand, ~8 lines)</summary>

```html
<form class="js-webhook-form" id="new_hook" action="/h3xx/frotz/-/hooks" accept-charset="UTF-8" method="post">
    <div>
        <input id="webhook-url" name="hook[url]" type="text" placeholder="http://example.com/trigger-ci.json" class="gl-form-input form-control is-valid" data-testid="form-url" aria-describedby="__BVID__4__BV_description_" autocomplete="off">
    </div>
    <div>
        <input autocomplete="new-password" class="form-control gl-form-input" type="password" name="hook[token]" id="hook_token">
    </div>
</form>
```

</details>

The browser **still** asks to save the token as password. 🤷 

Firefox test:

| URL Field | Token Field | Browser | KeepassXC-browser
| - | - | - | -
| Unmodified (no `type` no `autocomplete`) | `type=password autocomplete=new-password` | asks | asks
| `type=url autocomplete=off` | `type=password autocomplete=new-password` | asks | does not ask

This means the Mozilla docs appear to not describe how Firefox works, which is somewhat annoying.

In conclusion, I'm not sure we can get away from the browser asking to save the auth token in its password vault.

Interesting to note that when I set the token field as `autocomplete=new-password`, the Firefox pops down an option below the active field to generate a secure random password, but only when "ask to save passwords" is enabled.

**Comment by silverwind** [comment]
As per https://stackoverflow.com/questions/41217019, there appears to be no documented way to suppress these storage messages, so I'm fine with accepting this, despite this minor annoyance.

**Review by silverwind — APPROVED** [review]

**Comment by silverwind** [comment]
There may be another issue: Browser and password managers will sometimes "helpfully" auto-fill any plain `type=password` field. To prevent that, `autocomplete=new-password` is the current recommended solution that has the most widespread support. Let's add it.

File: templates/repo/migrate/gitea.tmpl [hunk 1]
```diff
@@ -20,7 +20,7 @@
 
 					<div class="inline field {{if .Err_Auth}}error{{end}}">
 						<label for="auth_token">{{.locale.Tr "access_token"}}</label>
-						<input id="auth_token" name="auth_token" value="{{.auth_token}}" {{if not .auth_token}} data-need-clear="true" {{end}}>
+						<input id="auth_token" name="auth_token" type="password" value="{{.auth_token}}" {{if not .auth_token}} data-need-clear="true" {{end}}>
```
Comment (silverwind) [inline templates/repo/migrate/gitea.tmpl]:
```suggestion
						<input id="auth_token" name="auth_token" type="password" autocomplete="new-password" value="{{.auth_token}}" {{if not .auth_token}} data-need-clear="true" {{end}}>
```

File: templates/repo/migrate/github.tmpl [hunk 2]
```diff
@@ -20,7 +20,7 @@
 
 					<div class="inline field {{if .Err_Auth}}error{{end}}">
 						<label for="auth_token">{{.locale.Tr "access_token"}}</label>
-						<input id="auth_token" name="auth_token" value="{{.auth_token}}" {{if not .auth_token}}data-need-clear="true"{{end}}>
+						<input id="auth_token" name="auth_token" type="password" value="{{.auth_token}}" {{if not .auth_token}}data-need-clear="true"{{end}}>
```
Comment (silverwind) [inline templates/repo/migrate/github.tmpl]:
```suggestion
						<input id="auth_token" name="auth_token" type="password" autocomplete="new-password" value="{{.auth_token}}" {{if not .auth_token}}data-need-clear="true"{{end}}>
```

File: templates/repo/migrate/gitlab.tmpl [hunk 3]
```diff
@@ -20,7 +20,7 @@
 
 					<div class="inline field {{if .Err_Auth}}error{{end}}">
 						<label for="auth_token">{{.locale.Tr "access_token"}}</label>
-						<input id="auth_token" name="auth_token" value="{{.auth_token}}" {{if not .auth_token}}data-need-clear="true"{{end}}>
+						<input id="auth_token" name="auth_token" type="password" value="{{.auth_token}}" {{if not .auth_token}}data-need-clear="true"{{end}}>
```
Comment (silverwind) [inline templates/repo/migrate/gitlab.tmpl]:
```suggestion
						<input id="auth_token" name="auth_token" type="password" autocomplete="new-password" value="{{.auth_token}}" {{if not .auth_token}}data-need-clear="true"{{end}}>
```

File: templates/repo/migrate/gogs.tmpl [hunk 4]
```diff
@@ -20,7 +20,7 @@
 
 					<div class="inline field {{if .Err_Auth}}error{{end}}">
 						<label for="auth_token">{{.locale.Tr "access_token"}}</label>
-						<input id="auth_token" name="auth_token" value="{{.auth_token}}" {{if not .auth_token}} data-need-clear="true" {{end}}>
+						<input id="auth_token" name="auth_token" type="password" value="{{.auth_token}}" {{if not .auth_token}} data-need-clear="true" {{end}}>
```
Comment (silverwind) [inline templates/repo/migrate/gogs.tmpl]:
```suggestion
						<input id="auth_token" name="auth_token" type="password" autocomplete="new-password" value="{{.auth_token}}" {{if not .auth_token}} data-need-clear="true" {{end}}>
```

**Review by silverwind — CHANGES_REQUESTED** [review]
Withdrawing

**Comment by h3xx** [comment]
Merged your suggestions, rebased off current `main`.

**Review by silverwind — APPROVED** [review]

**Review by wxiaoguang — APPROVED** [review]
If no other objection, let's take this change?

**Comment by silverwind** [comment]
.

**Comment by silverwind** [comment]
.

**Comment by techknowlogick** [comment]
🤖 🎺
